[documentation]: Release roadmap (#3809)
hyperledger/iroha: docs/source/release_roadmap.org

> #+TITLE: Recommended Road

Reviewer: ```suggestion
#+TITLE: Recommended Release Roadmap for IROHA v2
```

#+AUTHOR: Aleksandr Petrosyan
#+DATE: 2023-08-13

* State of the project

Iroha v2 is a capable ledger with notable implemented subroutines that

> allow it to be deployed successfully as a private blockchain ledger:

Reviewer 2: I have a feeling that some points sounds a bit subjective. It will be better if we can provide some quantitative measurements here.
Like "good performance characteristics" compared to what?
Or "is a capable ledger" does it have feature parity with it's competitors in the private blockchain context?

Author: Good point.

1. It has relatively good performance characteristics regarding
   throughput (post RC9)
2. It has good performance characteristics regarding transaction
   finality and latency (post RC19 also query latency).

> 3. It has support for Turing complete permission validation, able to

Reviewer: ```suggestion
3. It has support for Turing complete permission validation and is able to
```

   allows Iroha to enforce arbitrarily complex invariants on-chain. 

> 4. It has support for a small interpreted VM called the Iroha Special

Reviewer 3: Is it correct to call it an "interpreted VM"

Author: Clarified.

   Instructions which allow small-footprint transactions
5. It has support for Turing complete smartcontracts living in a
   sandbox, with a full array of events which can trigger these
   behaviours.
6. The ledger has three SDKs within the JVM, JavaScript and Python
   ecosystems, allowing the users to easily adapt the Iroha node to

>    their requirements. 

Author: Yes. I've added it as point 7.

The ledger has been demonstrated to work during the ADB CSD/RTGS
linkages project, and is slated to be used in two projects internally

> by Soramitsu which only utilise the private blockchain ledger

Reviewer: ```suggestion
by SORAMITSU which only utilise the private blockchain ledger
```

functionality.

The ledger can support un-sharded monolithic blockchains, with limited
export capability (unlike Iroha v1, v2 uses an internal representation
and doesn't use a database). Its consensus was designed with Byzantine

> fault tolerance in mind, however, the algorithm -- =sumeragi= has some

Reviewer 4: Should the name be capitalized?

Author: Fixed.

intrinsic flaws outlined in [[https://arxiv.org/pdf/1707.01873.pdf][an IBM Research article]], which is notable
because IBM research originally designed the B-Chain algorithm which

> serves as a prototype for =sumeragi=.

Reviewer 4: same as above

> intrinsic flaws outlined in [[https://arxiv.org/pdf/1707.01873.pdf][an IBM Research article]], which is notable

Reviewer 3: the paper doesn't mention any intrinsic flaws in B-Chain anywhere in the paper

Author: not of BChain. I've added quotes to make it clearer what I mean.

> because IBM research originally designed the B-Chain algorithm which

Reviewer 3: not sure what this means. Sisi Duan did this for her PhD thesis at UC Davis

The blockchain ledger is designed to support a plug-able (and
upgrade-able) permission validator modules, allowing the blockchain
invariants to evolve and change with the operation of the blockchain.
It is capable of resolving simple soft-forks, and has tooling to
quickly restart a blockchain by squashing all of the information and
history into a single genesis block. However, it does not retain
enough flexibility to be used as a single ledger for both private and
public blockchains. And the following issues have been identified.

1. The more efficient ISI-based language is not usable. The

>    expressions are not sufficiently expressive.

Reviewer: 🤯

   While this can be somewhat mitigated by the introduction of the
   DSL, the main expressiveness-boosting facilities of ISI are not
   usable, and thus only add overhead. A thorough exploration of a
   restructuring of the Iroha Instruction Set Architecture is needed
   and shall be published as an RFC in August 2023.

> 2. WASM is a sub-optimal medium for storage of executable logic. It

Reviewer 4: Thanks for pointing it out :-)

   alone is responsible for 80% of the complexity present in the code

>    base and around 95% of technical debt is related to some mitigation

Reviewer 2: Can you please elaborate on this numbers?

Author: `no_std`. Much of the dancing around in the `ffi` library has more to do with what is possible on WASM. The fact that the builds are not reproducible can technically be lumped in here as well, given that things like [reproducible central](https://reproducible-builds.org/docs/jvm/) exist for JVM languages.

>    of a shortcoming of WASM in the blockchain.

Reviewer 5: tl;dr it also limits SDK usability

referring to SDKs, a developer willing to use smart contracts is basically forced to write ones in Rust since compiling almost everything else to WASM is experimental/prototypical/highly challenging
moving that to the side of SDKs is an overkill at this point

Reviewer 5: ah, I see it basically mentioned later, nvm

Reviewer 2: @Mingela hmm, i don’t see how picking other runtime will stop limiting SDK devs. If we will pick lua then sevs have to write contracts in lua. If jvm them write contract in jvm based language and so on…
Can you explain it a bit further?

Author: Well, this will be expanded upon in the WASM-specific document. The main issue with WASM specifically is that it is tailored to a web-based use, which has some countervailing tendencies to what we see in general. There's no pressure to stabilise modules, hence no dynamic linkage. It is meant to transport low-level compiled instructions thus no or non-existent run-time support is present. WASM takes up more RAM. WASM must be compiled. WASM is not backwards compatible and WASM must be AOT compiled and then JIT-ted for the particular run-time.  

With a different standard we get a bit more freedom, because the different standard is designed for specific uses. The JVM is perhaps the closest point of comparison. It too has a limited number of supported languages that can target it, but a much wider coverage. Because most JVM languages assume a host-side garbage collector, one has automatic memory management without either the borrow checker, or having to embed GC. Java is smaller for the same source code, because more of the data is embedded in the runtime. 

The runtime itself can be large, but because of late binding the actual memory usage is smaller. Plus we don't have to come up with a dynamic linkage mechanism, we have one. We also have functional multithreading, and 64-bit support. Another important aspect for why JVM would be better is the guaranteed backwards compatibility. Java does not make breaking changes, and you don't have to keep an older version of a runtime in order to preserve compatibility. This means that instead of a complex system of host-side runtime selection, we can just upgrade to the newer version. 

Thus with Java we get a slightly less complex language for smartcontracts that can do slightly more, with the added benefit of being a huge ecosystem tailored to running application logic (and not DOM manipulation), that is significantly more mature. WASM is not the worst possible standard, but we can alleviate a lot of the problems. 

AND THEN, we get to interpreted languages, which have lightweight syntax, don't allow you to do what's not intended, can live concurrently and allow for a lot of transparency. Once I give you the WASM blob, you have a whole song and dance to verify that it belongs to the source code that I gave you. Most people wouldn't bother and that discourages audits. At the same time, saying that the smartcontract is an actual source snippet that is interpreted in the blockchain, tells you a lot more about the logic of the smartcontract and gives you a lot more trust. 

Because for an interpreted language we must expose the FFI bindings, we could then just say "you know what, we don't care, as long as it conforms to some very simple protocol, use Python if that rocks your boat". Then the language becomes an implementation detail for each blockchain, and we're not confined to _one_ choice.

   The advantages and disadvantages for various media storing Turing
   complete logic on chain must also be discussed and shall be
   published as an RFC by the end of August 2023.

> 3. Iroha lacks any facilities for correlating the consensus process

Reviewer: ```suggestion
3. Iroha lacks any facilities to correlate the consensus process
```

   with the blockchain data. This prevents Proof-of-work,

>    Proof-of-stake, slashing, governance and democracy from being

Reviewer: ```suggestion
   Proof-of-stake, slashing, governance, and democracy from being
```

   beyond 8 cores. This produces a set of system requirements that
   are unlikely to occur naturally in a public blockchain.

6. An Iroha network does not have a stable API. As such custom public
   blockchains are liable to frequently squash their block history or
   eschew upgrades.

7. Iroha doesn't have a distinct understanding of the blockchain
   uses. It caters to the private blockchains and as such has
   validation for IP addresses, and some logic to handle mint-able and
   non-mint-able assets as well as roles and permission tokens. This
   information is hard-coded and non-modular, leading to public
   blockchains missing vital functionality and having to commit their
   changes upstream, and thus leading to incompatibility between the
   Iroha APIs for a public and private blockchain.

8. Iroha has no facilities to extend the expression language. This
   dovetails into point 1, and leads to even less logic being encoded
   as an ISI or =Expression= and instead relying heavily on WASM.

9. WASM is only well-supported by Rust. While WASM is a
   general-purpose binary format, only a few languages support it

>    natively: =C++=, =Go=, =nim=, =zig= and =Rust=. =Java=, =Kotlin=

Reviewer 2: This languages cover pretty large share of used programming languages according to this [statistics](https://www.statista.com/statistics/793628/worldwide-developer-survey-most-used-languages/). Also i would noticed that this list is not [full](https://github.com/appcypher/awesome-wasm-langs).

Author: For one, this list already contains a half-truth: Rust only supports wasm32-unknown-unknown as a [tier 2 target](https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2) and wasm64 is a tier 3. 

Second, Java does not officially support it as a target, instead providing a "use-at-your-own-risk" workaround (you can see that in your second link), corroborated by [official WASM docs](https://webassembly.org/getting-started/developers-guide/). 

Thirdly, I don't think that that statistic supports your claim. In fact, the first item that technically qualifies (see second point) is Java, which ranks 7-th. So at best, we cover only 30% of developers worldwide, unless we start shipping v8 and allowing people to write financial code in JavaScript (not saying it couldn't be done, but it's like hammring a nail with a screwdriver). 

Fourth, awesome-* listicles are not reliable sources of information. One such listicle listed emacs as a terminal editor, gave Vim and Neovim one entry (despite the two projects being extremely different) and listed VSCode and VSCodium as two separate entries, despite the former being a proprietary version of the latter.

Author: I will admit that the phrasing makes it look like this list is meant to be exhaustive. I'll fix that.

Reviewer 5: Unfortunately, most of stuff mentioned as 'supporting WASM' provide hardly usable, very limited, experimental prototypes. So it rather indicates whether some attempts were made in that direction at all, not more. Agree with Aleksandr on that.

   and a few other languages have experimental support with little in
   terms of either interest or effort.

   Note that all of the above languages are ahead-of-time compiled

>    systems programming languages which require low-level understanding

Reviewer 2: Nitpicking but i wouldn't call `Java` or `Kotlin` system programming language.
This might indicate conformation bias regarding usage of `WASM`.

Author: Will fix. I included Java and Kotlin _post factum_ and forgot to propagate the changes.

more changes needed to be implemented downstream for all engineers. It
can be somewhat mitigated by separating releases into /stable/ and
/LTS/, however the brunt of the API changes for making Iroha v2 double
as both a private and public ledger will have to be carried by the
downstream users. Those who can choose a different technology will do
so, those who cannot will be less productive.

As such, it is recommended to all key stakeholders of the Iroha v2
project, to consider the release of version 2.0.0 before any of the
public blockchain features have been added. Based on this
recommendation a further recommendation of a sequence of steps shall
be considered, based on a few key decisions.

* Scenario 0: Status quo

In this case the following is true:
- The ISI and Expression language is kept as is,
- The WASM infrastructure is continued to be built as per the original plan
- Public blockchain features are added and until the previously agreed
  upon set of features is implemented (TODO: link Zenhub), no v2.0.0
  release is carried out.
- =sumeragi= is kept as the only supported consensus protocol and is
  supplemented with features necessary to support uses in a public

>   blockchain, but no concerns outlined in the IBM research article is

Reviewer 4: "are addressed"?

Author: English plurals are weird.

prove more ergonomic than e.g. Parity Substrate, it will not be
competitive with Solana. Given lack of interest in interoperability
with SoLang, Iroha is likely to alienate users.

Having gone through a few iterations and actually being deployed in a
public blockchain the necessity to accumulate backwards-incompatible
versions of the WASM virtual machine will necessitate a further
reduction in throughput, and require further mitigation strategies to
cope with the large memory requirements. Further, at present the
virtual machine is statically linked. Adding an extra layer of dynamic
linkage will result in a further overall reduction in efficiency given
that permission validation (and the Executor RFC), are to be
implemented and thus all ISI execution will go through the WASM
sub-process.

Furthermore, any change in the API will be accompanied by an extra
copy of the =iroha_data_model= crate embedded into the Iroha runtime
in addition to the copies that will occupy space in the Executor. 

As such a monthly release cadence is the preferred solution as it
allowed for a stable communication channel with the SDKs and the
community, further boosting the profile of the project.

> * Scenario 2: ISA refactor only

Reviewer 5: did `Scenario 1` get missed on purpose

Author: Yes. I forgot to update the numbers.

data model which affect the communication of the Executor to the host
node remain frozen and thus don't need to be updated. Because
instructions no longer enforce any of the blockchain invariants, this
is not restrictive and a natural method of extension is provided by
extending the Executor with custom expressions.

No further benefits.

* Scenario 3: ISA refactor and change of smartcontract language

Depending on which smartcontract language is realised, the technical
limitations incurred by usage of WASM can either be eliminated or
substantially reduced. The precise character of the reductions shall
be outlined in the RFC. But the short version follows: standards with
guaranteed backwards compatibility, such as Java, native C-ABI
libraries, and an embedded scripting language all eliminate the need
for hosting multiple versions of multiple machines for validation.
This greatly simplifies deployment of new nodes, attracting more users
in the public blockchain space, as well as reducing the operational
costs to private blockchains.

Removal of WASM, will definitely eliminate the need to implement
shared-memory dynamic linkage, and instead use the standard-native

> protocol. C-ABI dynamic linkage is implemented by the operating

Reviewer 2: I think providing `C-ABI` compatible layer for iroha already proved to be non-trivial task and still ongoing effort.

Author: Actually, providing a C-ABI compatible layer proved to be a trivial task. What was a problem was making it suitable for smart contract use. @mversic can say more.

system, resulting in the best outcome. For Java some bridging code may

> be needed. For =lua= and Python, the dynamic linkage is an

Reviewer 4: I wonder if I should point out the existence of [LuaJIT](https://luajit.org/), which is heavily optimized for `x86` and `x86-64` platforms? If we're talking about Lua, we must consider maximizing its performance.

Author: Remind me of that when I create the WASM-away RFC.

implementation-specific detail of the interpreter, and only minor
adjustments may be needed to use the exported functions via a
Foreign-function interface. For the record the same work must be done
for WASM to reduce the size of smartcontracts, which is not a
consideration on a private blockchain, but might become a problem in a
public use-case.

Depending on the specific choice, the popularity might either be
improved substantially or be at least as good as with WASM. C-ABI
dynamic linkage offers the widest compatibility, as it is a ubiquitous
operating-system-guaranteed interface. One could, in principle write
smartcontracts in regular =x86= assembly to offer the best size
optimisations. This can be combined with a scripting language also
being native supported on-chain, because languages like =lua= and
Python offer support for native linkage as part of the standard
interpreter. The JVM would offer a wider range of compatible programs,
and though the full runtime library might be in excess of a few
hundred megabytes, this cost must be paid once up-front and provide

> compatibility with a wide range of languages

Reviewer: ```suggestion
compatibility with a [wide range of languages](https://en.wikipedia.org/wiki/List_of_JVM_languages).
```

Author: @WRRicht3r This is an org document, the link syntax is slightly different: 
```org
[[link text][link destination]]
```

> [[https://en.wikipedia.org/wiki/List_of_JVM_languages]].

Reviewer: ```suggestion

```
